buildSrc/src/main/kotlin/accessors/accessors.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package accessors
+
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.reporting.ReportingExtension
+
+import org.gradle.kotlin.dsl.*
+
+
+val Project.java
```

Comment:
I think this is exactly what people will start doing, making the point about "we don't want to offer an accessor if the plugin is not applied" moot. So instead of making everyone write this, why not generate accessors for everything on the classpath? We can still refine this later by improving the plugin development plugin, but at least people won't have to go through this boilerplate.

Reply:
We're going to implement accessors for declared plugin dependencies as part of Kotlin DSL 1.0 (see gradle/kotlin-dsl#673).

> why not generate accessors for everything on the classpath?

And use which process to infer the data required for code generation (the project schema as we call it)? 

E.g. what would be the implementation of the function `fun projectSchemaFor(classPath: ClassPath): ProjectSchema<*>`?

Reply:
The tooling api supports classpath injection, so one could create a dummy build with that classpath and then do one invocation per plugin on the classpath to find out what it exposes. Not terribly efficient of course. A more explicit mechanism would obviously be better :)